src/P3Scheme.jl

```diff
@@ -667,4 +767,464 @@ function D_m(p3::PSP3, q::FT, N::FT, ρ_r::FT, F_r::FT) where {FT}
+    # TODO make this more accurate for non-spherical particles 
+    return π * (D1 / 2 + D2 / 2)^2
+end
+
+"""
+    N′colliding(type, D, q, N)
+
+ - type - defines what is colliding with ice ("rain" or "cloud")
+ - D - maximum dimension 
+ - q - mass mixing ratio 
+ - N - number mixing ratio 
+
+ Returns the corresponding distribution of particles depending on type
+"""
+function N′colliding(type::String, D::FT, q::FT, N::FT) where {FT}
+    if type == "rain"
+        return N′rain(D, q, N)
+    elseif type == "cloud"
+        return N′cloud(D, q, N)
+    else
+        throw("Undefined Type of Collisions")
+    end
+end
+
+"""
+    vel_diff(type, D, Dᵢ, p3, Chen2022, ρ_a, F_r, th)
+
+ - type - defines what is colliding with ice ("rain" or "cloud")
+ - D - maximum dimension of colliding particle 
+ - Dᵢ - maximum dimension of ice particle 
+ - p3 - a struct containing P3 parameters
+ - Chen2022 - a struct containing Chen 2022 velocity parameters 
+ - ρ_a - density of air 
+ - F_r - rime mass fraction (q_rim/ q_i)
+ - th - thresholds as calculated by thresholds()
+
+ Returns the corresponding velocity difference of colliding particles depending on type
+"""
+function vel_diff(
+    type::String,
+    D::FT,
+    Dᵢ::FT,
+    p3::PSP3,
+    Chen2022::CMP.Chen2022VelType,
+    ρ_a::FT,
+    F_r::FT,
+    th,
+) where {FT}
+    if type == "rain"
+        return abs(
+            velocity_chen(Dᵢ, p3, Chen2022.snow_ice, ρ_a, F_r, th) -
+            velocity_chen(D, Chen2022.rain, ρ_a),
+        )
+    elseif type == "cloud"
+        return abs(velocity_chen(Dᵢ, p3, Chen2022.snow_ice, ρ_a, F_r, th))
+    end
+end
+
+"""
+    ice_collisions(type, p3, Chen2022, ρ_a, F_r, qᵣ, qᵢ, Nᵣ, Nᵢ, ρ, ρ_r, E_ri)
+
+ - type - defines what is colliding with the ice ("cloud" or "rain")
+ - p3 - a struct with P3 scheme parameters 
+ - Chen2022 - a struct with terminal velocity parameters as in Chen (2022) 
+ - qᵢ - mass mixing ratio of ice 
+ - Nᵢ - number mixing ratio of ice 
+ - q_c - mass mixing ratio of colliding species 
+ - N_c - number mixing ratio of colliding species
+ - ρ_a - density of air 
+ - F_r - rime mass fraction (q_rim/ q_i) 
+ - ρ_r - rime density (q_rim/B_rim) 
+ - T - temperature (in K)
+ - E_ci - collision efficiency between ice and colliding species 
+
+ Returns the rate of collisions between cloud ice and rain 
+ Equivalent to the measure of QRCOL in Morrison and Mildbrandt (2015)
+"""
+function ice_collisions(
+    type::String,
+    p3::PSP3,
+    Chen2022::CMP.Chen2022VelType,
+    qᵢ::FT,
+    Nᵢ::FT,
+    q_c::FT,
+    N_c::FT,
+    ρ_a::FT,
+    F_r::FT,
+    ρ_r::FT,
+    T::FT,
+    E_ci = FT(1),
+) where {FT}
+    ρ_w = FT(1000)      # TO DO: add this as density_liquid_water from ClimaParams
+    th = thresholds(p3, ρ_r, F_r)
+    (λ, N_0) = distribution_parameter_solver(p3, qᵢ, Nᵢ, ρ_r, F_r)
+    (colliding_bound, ice_bound) =
+        integration_bounds(type, p3, eps(FT), λ, N_0, Nᵢ, q_c, N_c)
+
+    if T > 273.15
```

Comment:
This should be `T_freeze` from ClimaParams. And agreed that `rho_w` should be taken from ClimaParams also. We can add them to the `p3` parameters and then just grab them from the struct here. Since we are already passing in the `p3` parameters struct. Lmk if you need help with that